As a developer, I'm encountering an NPE in staging when running a test HEFS evaluation

The declaration is below (data is already in Google Drive for VLab, most likely). The exception:

```
2025-05-30T17:12:52.296+0000  [pool-75-thread-1] INFO EvaluationService - The evaluation failed with the following stack trace: 
wres.pipeline.InternalWresException: Could not complete project execution
	at wres.pipeline.Evaluator.evaluate(Evaluator.java:407)
	at wres.pipeline.Evaluator.evaluate(Evaluator.java:179)
	at wres.Functions.evaluate(Functions.java:135)
	at wres.server.EvaluationService.lambda$startEvaluation$2(EvaluationService.java:695)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: wres.pipeline.WresProcessingException: Encountered an error while processing evaluation 'PGOpMr-7vlJ9COiztEoED86ypqI': 
	at wres.pipeline.Evaluator.evaluate(Evaluator.java:844)
	at wres.pipeline.Evaluator.evaluate(Evaluator.java:378)
	... 7 common frames omitted
Caused by: wres.io.ingesting.IngestException: An ingest task could not be completed.
	at wres.io.ingesting.SourceLoader.load(SourceLoader.java:165)
	at wres.pipeline.Evaluator.evaluate(Evaluator.java:600)
	... 8 common frames omitted
Caused by: java.util.concurrent.CompletionException: wres.io.ingesting.IngestException: Failed to get ingest results.
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:649)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1773)
	... 3 common frames omitted
Caused by: wres.io.ingesting.IngestException: Failed to get ingest results.
	at wres.io.ingesting.database.DatabaseTimeSeriesIngester.ingest(DatabaseTimeSeriesIngester.java:290)
	at wres.io.ingesting.SourceLoader.lambda$readAndIngestData$4(SourceLoader.java:398)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646)
	... 5 common frames omitted
Caused by: java.util.concurrent.ExecutionException: java.lang.NullPointerException: Cannot invoke "wres.io.database.details.SourceDetails.getId()" because the return value of "wres.io.database.caching.DataSources.getSource(String)" is null
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
	at wres.io.ingesting.database.DatabaseTimeSeriesIngester.ingest(DatabaseTimeSeriesIngester.java:271)
	... 7 common frames omitted
Caused by: java.lang.NullPointerException: Cannot invoke "wres.io.database.details.SourceDetails.getId()" because the return value of "wres.io.database.caching.DataSources.getSource(String)" is null
	at wres.io.database.caching.DataSources.getSourceId(DataSources.java:291)
	at wres.io.ingesting.database.DatabaseTimeSeriesIngester.getSurrogateKey(DatabaseTimeSeriesIngester.java:1213)
	at wres.io.ingesting.database.DatabaseTimeSeriesIngester.finalizeExistingSource(DatabaseTimeSeriesIngester.java:1058)
	at wres.io.ingesting.database.DatabaseTimeSeriesIngester.ingestEnsembleTimeSeries(DatabaseTimeSeriesIngester.java:702)
	at wres.io.ingesting.database.DatabaseTimeSeriesIngester.ingestEnsembleTimeSeriesWithRetries(DatabaseTimeSeriesIngester.java:543)
	at wres.io.ingesting.database.DatabaseTimeSeriesIngester.lambda$ingest$1(DatabaseTimeSeriesIngester.java:256)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	... 3 common frames omitted
```

NPEs are always bad. I'll begin by seeing if its readily reproducible.

Hank

======================

```yaml
label: BV Streamflow - Hank Test
observed:
  label: OBS Streamflow
  sources:
  - uri: ABRN1_QME.xml
  variable:
    name: QME
  feature_authority: nws lid
  time_scale:
    function: mean
    period: 24
    unit: hours
  type: observations
predicted:
  label: HEFS Streamflow
  sources:
  - uri: ABRN1_Streamflow_HEFS_Baseline_Validation.tgz
  variable:
    name: SQIN
  time_scale:
    function: mean
    period: 1
    unit: minutes
  type: ensemble forecasts
baseline:
  label: ESP Streamflow
  sources:
  - uri: ABRN1_Streamflow_ESP_Baseline_Validation.tgz
  variable:
    name: SQIN
  time_scale:
    function: mean
    period: 1
    unit: minutes
  type: ensemble forecasts
unit: ft3/s
lead_time_pools:
  period: 24
  frequency: 24
  unit: hours
cross_pair: exact
lead_times:
  minimum: 18
  maximum: 720
  unit: hours
season:
  minimum_day: 1
  minimum_month: 1
  maximum_day: 31
  maximum_month: 12
time_scale:
  function: mean
  period: 24
  unit: hours
pair_frequency:
  period: 24
  unit: hours
rescale_lenience: none
features:
- observed: ABRN1
  predicted: ABRN1
  baseline: ABRN1
probability_thresholds:
- values:
  - 0.01
  - 0.1
  - 0.5
  - 0.9
  - 0.95
  - 0.99
  - 0.995
  - 0.999
  operator: greater
  apply_to: observed
metrics:
- name: brier score
- name: brier skill score
- name: pearson correlation coefficient
- name: quantile quantile diagram
- name: bias fraction
- name: continuous ranked probability score
- name: continuous ranked probability skill score
- name: mean absolute error
- name: mean error
- name: root mean square error
- name: box plot of errors by forecast value
- name: box plot of errors by observed value
- name: rank histogram
- name: relative operating characteristic diagram
- name: relative operating characteristic score
- name: reliability diagram
- name: sample size
duration_format: hours
decimal_format: '0.000'
output_formats:
- format: csv2
- format: png
  width: 800
  height: 600
  orientation: lead threshold
```

--- a/wres-io/src/wres/io/database/caching/DataSources.java
+++ b/wres-io/src/wres/io/database/caching/DataSources.java
@@ -16,7 +16,8 @@ import wres.io.database.DataScripter;
 import wres.io.database.Database;
 
 /**
- * Caches information about the source of forecast and observation data
+ * Caches information about data sources.
+ *
  * @author James Brown
  * @author Christopher Tubbs
  */
@@ -58,7 +59,7 @@ public class DataSources
     }
 
     /**
-     * Creates an instance
+     * Creates an instance.
      * @param database the database
      * @param data the data source
      * @throws NullPointerException if either input is null
@@ -97,7 +98,7 @@ public class DataSources
     /**
      * Gets the source details corresponding to the source hash.
      * @param hash the hash
-     * @return the source details
+     * @return the source details, possibly null
      * @throws SQLException if the source could not be found
      * @throws NullPointerException if the input is null
      */
@@ -113,7 +114,7 @@ public class DataSources
     /**
      * Gets the source details from the corresponding source identifier.
      * @param id the source identifier
-     * @return the source details
+     * @return the source details, possibly null
      * @throws NullPointerException if the input is null
      */
 
@@ -125,9 +126,9 @@ public class DataSources
     }
 
     /**
-     * Gets the ID of source metadata from the instanced cache based on identity
+     * Gets the ID of source metadata from the instanced cache based on identity.
      * @param key the hash code for the source file
-     * @return The ID of the source in the database
+     * @return The ID of the source in the database, possibly null
      * @throws SQLException Thrown when interaction with the database failed
      * @throws NullPointerException if the input is null
      */
@@ -157,7 +158,7 @@ public class DataSources
     /**
      * Gets an existing source based on the source identifier and adds to the cache if required.
      * @param id the source identifier
-     * @return the source details
+     * @return the source details, possible null
      * @throws SQLException if an error occurs when inspecting the database
      * @throws NullPointerException if the input is null
      */
@@ -221,7 +222,7 @@ public class DataSources
     /**
      * Gets an existing source based on the source hash and adds to the cache if required.
      * @param hash the source hash
-     * @return the source details
+     * @return the source details, possibly null
      * @throws SQLException if an error occurs when inspecting the database
      * @throws NullPointerException if the input is null
      */
@@ -280,15 +281,22 @@ public class DataSources
     /**
      * Gets the source identifier corresponding to the source hash.
      * @param hash the hash
-     * @return the source identifier
+     * @return the source identifier, possibly null
      * @throws SQLException if an error was encountered when checking the database
      */
 
     public Long getSourceId( String hash )
             throws SQLException
     {
-        return this.getSource( hash )
-                   .getId();
+        Long sourceId = null;
+        SourceDetails sourceDetails = this.getSource( hash );
+
+        if( Objects.nonNull( sourceDetails ) )
+        {
+            sourceId = sourceDetails.getId();
+        }
+
+        return sourceId;
     }
 
     /**

--- a/wres-io/src/wres/io/ingesting/database/DatabaseTimeSeriesIngester.java
+++ b/wres-io/src/wres/io/ingesting/database/DatabaseTimeSeriesIngester.java
@@ -1059,7 +1059,8 @@ public class DatabaseTimeSeriesIngester implements TimeSeriesIngester
 
             // First, try to safely remove it, which requires an exclusive lock
             boolean removed = false;
-            if ( this.lockSource( source, null, this.getLockManager() ) )
+            if ( surrogateKey != KEY_NOT_FOUND
+                 && this.lockSource( source, null, this.getLockManager() ) )
             {
                 try
                 {

